clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/view/OpenstackSecurityGroupProvider.groovy

```diff
@@ -0,0 +1,100 @@
+
+import com.netflix.spinnaker.clouddriver.model.SecurityGroupProvider
+import com.netflix.spinnaker.clouddriver.openstack.OpenstackCloudProvider
+import com.netflix.spinnaker.clouddriver.openstack.model.OpenstackSecurityGroup
+import com.netflix.spinnaker.clouddriver.openstack.security.OpenstackNamedAccountCredentials
+import com.netflix.spinnaker.clouddriver.security.AccountCredentialsProvider
+import groovy.util.logging.Slf4j
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.stereotype.Component
+
+@Slf4j
+@Component
+class OpenstackSecurityGroupProvider implements SecurityGroupProvider<OpenstackSecurityGroup> {
+
+  final AccountCredentialsProvider accountCredentialsProvider
+
+  @Autowired
+  OpenstackSecurityGroupProvider(AccountCredentialsProvider accountCredentialsProvider) {
+    this.accountCredentialsProvider = accountCredentialsProvider
+  }
+
+  Set<OpenstackSecurityGroup> getSecurityGroups() {
+    // TODO This is not thread safe and not cached
+    // This just proves the provider is working once it has a data set.
+    Set<OpenstackSecurityGroup> securityGroups = []
+    accountCredentialsProvider.all.each {account ->
+      if (account instanceof OpenstackNamedAccountCredentials) {
+        def provider = ((OpenstackNamedAccountCredentials) account).credentials.provider
+
+        provider.allRegions.each { region ->
+            securityGroups.addAll(provider.collect { p ->
+            p.getSecurityGroups(region)
+          }.flatten().collect{ sg ->
+            new OpenstackSecurityGroup(id: sg.id,
+              accountName: account.name,
+              name: sg.name,
+              description: sg.description,
+              region: region,
+              inboundRules: []) // TODO
+          })
+        }
+      }
+    }
+
+    securityGroups
+  }
+
+  @Override
+  String getType() {
+    return OpenstackCloudProvider.ID
+  }
+
+  @Override
+  Set<OpenstackSecurityGroup> getAll(boolean includeRules) {
+    getSecurityGroups()
+  }
+
+  @Override
+  Set<OpenstackSecurityGroup> getAllByRegion(boolean includeRules, String region) {
+    getSecurityGroups().findAll { it.region == region }
+  }
+
+  @Override
+  Set<OpenstackSecurityGroup> getAllByAccount(boolean includeRules, String account) {
+    getSecurityGroups().findAll { it.accountName == account }
+  }
+
+  @Override
+  Set<OpenstackSecurityGroup> getAllByAccountAndName(boolean includeRules, String account, String name) {
+    getAllByAccount(includeRules, account).findAll { it.name == name }
+  }
+
+  @Override
+  Set<OpenstackSecurityGroup> getAllByAccountAndRegion(boolean includeRule, String account, String region) {
+    getAllByAccount(includeRule, account).findAll { it.region == region }
+  }
+
+  @Override
+  OpenstackSecurityGroup get(String account, String region, String name, String vpcId) {
+    getAllByAccountAndRegion(true, account, region).findAll { it.name == name }[0]
```

Comment:
The findResult method will iterate through the collection and find the first non-null result.